### ISSUE TITLE
4.2 build system packaging, release, and dependency management issues

### ISSUE THREAD

**Issue by mcv21** [irods/irods#3277]
Sorry for a somewhat discursive bug report; I would have sent this to a mailing list for irods developers if there was one. At the end of the discussion, I've tried to note some specific issues as sub-issue.

The "externals" of irods 4.2.x builds a number of build-dependency packages, most of which become runtime dependencies of the resulting packages:

* irods-externals-libarchive3.1.2-0
* irods-externals-avro1.7.7-0
* irods-externals-boost1.60.0-0
* irods-externals-clang-runtime3.8-0
* irods-externals-jansson2.7-0
* irods-externals-zeromq4-14.1.3-0

Notably, this means an entire C++ runtime is being installed in parallel to the system's one.

There is, inevitably, a tension between supporting a range of distributions, particularly older ones like CentOS 6 / Ubuntu 12, and being able to use critical new features of libraries. In the case of libraries that distributions haven't picked up at all yet, providing a package of the relevant library is clearly necessary, including versions built on the older distributions.

The cost of requiring the use of libraries that are unsupported by distributions is substantially that someone has to bear that support burden. This means providing a clean upgrade path, and, more critically, doing the security support and back-porting of fixes that distributions typically do. With a few small libraries this load is not too onerous, but as the number and complexity of these extra-distribution libraries increase, this becomes a substantial burden of work that may be urgent and security-critical for production infrastructure.

There are two aggravating factors with the external dependencies currently required with irods 4.2: firstly, some of the libraries are very large and complex (boost, the c++ runtime); secondly, some of these libraries are not proper upstream releases, but rather arbitrary commits - sometimes particular commits with local patches on top. In the event of a vulnerability, we might reasonable expect upstream to provide patches that apply cleanly to their supported releases, but there's no particular guarantee that such patches would apply to a particular commit from between releases, never mind one that has been further modified.

These packages are far from [lintian](https://lintian.debian.org/)-clean, too; while that's understandable in packages that are essentially conveniences for a build system, it's less good in things that are going to be installed on production systems.

My other area of concern is the reproducibility of the build - installing the prerequisites and building "externals" involves a lot of downloading objects from the network; over https, but with no further integrity checking beyond that (e.g. with the ruby gems). Normally you would expect depencies expressed as "release X of software Y", mostly with X such that they can be satisfied from the distributions used in production. Currently some of them are essentially HEAD from a point in time, sometimes with local patches on top. My comments about duplicating the work of distributions apply here, but there is also the need for a particular version of irods to be fixed in terms of the versions of its "externals" - otherwise you might get a different artifact resulting from the build of ostensibly the same irods source. Our users are scientists, and [the reproducibility of software in science](http://arxiv.org/abs/1608.06897) is a live concern.

In summary, I would suggest the following:

- [ ] Use distribution versions of libraries wherever possible
- [ ] Security support for all external libraries must be available and suitably resourced for the support lifetime of the iRods release that depends upon them
- [ ] External library packages should be based on stable upstream releases
- [ ] Packages should comply with distributions' packaging policies (i.e. be lintian-clean)
- [ ] A given irods release should have clearly-versioned external library source and binary packages, with checksums

**Comment by Pete-C** [comment]
I would like to add that over the past few months, RENCI have been very helpful in responding to bug reports, providing assistance and ensuring resolution. Going forward, we are very happy to walk through our current pinch points with local upgrades, local development against iRODS (both internally and with 3rd parties) and package management. 
In particular, Matthew has been a long time Debian developer and is in an excellent position to help us with our downstream management and integration. If we can help, we'd be happy to engage wherever would be considered relevant
Pete 

    On Friday, 9 September 2016, 14:05, Matthew Vernon <notifications@github.com> wrote:
 

 Sorry for a somewhat discursive bug report; I would have sent this to a mailing list for irods developers if there was one. At the end of the discussion, I've tried to note some specific issues as sub-issue.The "externals" of irods 4.2.x builds a number of build-dependency packages, most of which become runtime dependencies of the resulting packages:   
   - irods-externals-libarchive3.1.2-0
   - irods-externals-avro1.7.7-0
   - irods-externals-boost1.60.0-0
   - irods-externals-clang-runtime3.8-0
   - irods-externals-jansson2.7-0
   - irods-externals-zeromq4-14.1.3-0
Notably, this means an entire C++ runtime is being installed in parallel to the system's one.There is, inevitably, a tension between supporting a range of distributions, particularly older ones like CentOS 6 / Ubuntu 12, and being able to use critical new features of libraries. In the case of libraries that distributions haven't picked up at all yet, providing a package of the relevant library is clearly necessary, including versions built on the older distributions.The cost of requiring the use of libraries that are unsupported by distributions is substantially that someone has to bear that support burden. This means providing a clean upgrade path, and, more critically, doing the security support and back-porting of fixes that distributions typically do. With a few small libraries this load is not too onerous, but as the number and complexity of these extra-distribution libraries increase, this becomes a substantial burden of work that may be urgent and security-critical for production infrastructure.There are two aggravating factors with the external dependencies currently required with irods 4.2: firstly, some of the libraries are very large and complex (boost, the c++ runtime); secondly, some of these libraries are not proper upstream releases, but rather arbitrary commits - sometimes particular commits with local patches on top. In the event of a vulnerability, we might reasonable expect upstream to provide patches that apply cleanly to their supported releases, but there's no particular guarantee that such patches would apply to a particular commit from between releases, never mind one that has been further modified.These packages are far from lintian-clean, too; while that's understandable in packages that are essentially conveniences for a build system, it's less good in things that are going to be installed on production systems.My other area of concern is the reproducibility of the build - installing the prerequisites and building "externals" involves a lot of downloading objects from the network; over https, but with no further integrity checking beyond that (e.g. with the ruby gems). Normally you would expect depencies expressed as "release X of software Y", mostly with X such that they can be satisfied from the distributions used in production. Currently some of them are essentially HEAD from a point in time, sometimes with local patches on top. My comments about duplicating the work of distributions apply here, but there is also the need for a particular version of irods to be fixed in terms of the versions of its "externals" - otherwise you might get a different artifact resulting from the build of ostensibly the same irods source. Our users are scientists, and the reproducibility of software in science is a live concern.In summary, I would suggest the following:   
   -  Use distribution versions of libraries wherever possible
   -  Security support for all external libraries must be available and suitably resourced for the support lifetime of the iRods release that depends upon them
   -  External library packages should be based on stable upstream releases
   -  Packages should comply with distributions' packaging policies (i.e. be lintian-clean)
   -  A given irods release should have clearly-versioned external library source and binary packages, with checksums
